pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+	mac, err := m.getter.GetMACByInterfaceNumber(ifaceNum)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface number by MAC %q: %w", mac, err)
+	}
+
+	links, err := m.rpdb.LinkList()
+	if err != nil {
+		return -1, fmt.Errorf("failed to list links: %w", err)
+	}
+
+	var link netlink.Link
+	for _, l := range links {
+		if l.Attrs().HardwareAddr.String() == mac {
+			link = l
+			break
+		}
+	}
+
+	return link.Attrs().Index, nil
+}
+
+func (m *migrator) copyRoutes(routes []netlink.Route, from, to int) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	// In ENI mode, we only expect two rules:
+	//   1) Link scoped route with a gateway IP
+	//   2) Default route via gateway IP
+	// We need to add the link-local scope route to the gateway first, then
+	// routes that depend on that as a next-hop later. If we didn't do this,
+	// then the kernel would complain with "Error: Nexthop has invalid
+	// gateway." with an errno of ENETUNREACH.
+	for _, r := range routes {
+		if r.Scope == netlink.SCOPE_LINK {
+			route := r
+			route.Table = to
+			if err := m.rpdb.RouteReplace(&route); err != nil {
+				return revertStack, fmt.Errorf("unable to replace link scoped route under table ID: %w", err)
+			}
+
+			revertStack.Push(func() error {
+				if err := m.rpdb.RouteDel(&route); err != nil {
+					return fmt.Errorf("failed to revert route upsert: %w", err)
+				}
+				return nil
+			})
+		}
+	}
+
+	for _, r := range routes {
+		if r.Scope == netlink.SCOPE_LINK {
+			// Skip over these because we already upserted it above.
+			continue
+		}
+
+		route := r
+		route.Table = to
+		if err := m.rpdb.RouteReplace(&route); err != nil {
+			return revertStack, fmt.Errorf("unable to replace route under table ID: %w", err)
+		}
+
+		revertStack.Push(func() error {
+			if err := m.rpdb.RouteDel(&route); err != nil {
+				return fmt.Errorf("failed to revert route upsert: %w", err)
+			}
+			return nil
+		})
+	}
+
+	return revertStack, nil
+}
+
+func (m *migrator) createNewRule(
+	rule netlink.Rule,
+	fromPrio, oldTable, toPrio, newTable int,
+) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	r := rule
+	r.Priority = toPrio
+	r.Table = newTable
+	if err := m.rpdb.RuleAdd(&r); err != nil {
+		return revertStack, fmt.Errorf("unable to add new rule: %w", err)
+	}
+
+	revertStack.Push(func() error {
+		if err := m.rpdb.RuleDel(&r); err != nil {
+			return fmt.Errorf("failed to revert rule insert: %w", err)
+		}
+		return nil
+	})
+
+	return revertStack, nil
+}
+
+func (m *migrator) deleteOldRule(rule netlink.Rule) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	if err := m.rpdb.RuleDel(&rule); err != nil {
+		return revertStack, fmt.Errorf("unable to delete old rule: %w", err)
+	}
+
+	revertStack.Push(func() error {
+		if err := m.rpdb.RuleAdd(&rule); err != nil {
+			return fmt.Errorf("failed to revert rule delete: %w", err)
+		}
+		return nil
+	})
+
+	return revertStack, nil
+}
```

Comment:
I don't think we need the revert logic here any more since it'd be ignored in the caller?